### ISSUE TITLE
[BUG] cuDF Deserialization with ucp buffer_region has incorrect dtype

### ISSUE THREAD

**Issue by quasiben** [rapidsai/cudf#2242]
cc @mrocklin @kkraus14 

When using cuDF with UCX the deserialization scheme fails.  Below is a small example of the failure.

UCX will create a buffer to a cuda pointer for the DeviceNDArray from the serialized cuDF Buffer.  This buffer has type and cuda information like the following:

```
In [8]: print(buffer_region.__cuda_array_interface__)
{'shape': (3,), 'typestr': '<i8', 'descr': [('', '<i8')], 'data': (140214108423168, False), 'version': 0}

In [9]: print(buffer_region.typestr)
<i8
```

However, as the ucx buffer is sent through the deserialization machinery it arrives here:
https://github.com/rapidsai/cudf/blob/b7b22feee1f175681eeffc1397be21d6731c84cd/python/cudf/cudf/dataframe/buffer.py#L41

At this point rmm takes over and the following code is called:
https://github.com/rapidsai/rmm/blob/branch-0.9/python/librmm_cffi/wrapper.py#L262-L265

```python
                if not isinstance(obj, np.ndarray):
                    obj = np.asarray(obj)
                cuda.devicearray.sentry_contiguous(obj)
                devobj = self.device_array_like(obj, stream=stream)
```

the UCX buffer is not an ndarray so now.  When we call `np.asarray` on the UCX buffer the dtype is down cast to uint8. 

I think, if we add:

```python
        if hasattr(obj, '__cuda_array_interface__'):
            return cuda.as_cuda_array(obj), False
``` 

to `librmm_cffi/wrapper.py` then that should resolve the issue

## Reproducible Failure:

A partially contrived example of the failure when a UCP buffer of a cudf.Series serialized  DeviceNDArray is then deserialized. 

```
import functools
import numpy as np
import cudf
import distributed
import ucp

serialize = distributed.protocol.serialize
cuda_serialize = functools.partial(serialize, serializers=["cuda"])
df = cudf.Series([1, 2, 3])
print(df.dtype)
assert df.dtype == np.int64
header, frames = cuda_serialize(df)

mem = frames[-2] # DeviceNDArray

buffer_region = ucp.BufferRegion()
buffer_region.populate_cuda_ptr(mem)
print(buffer_region.__cuda_array_interface__)
print(buffer_region.typestr)
assert buffer_region.typestr == np.dtype('<i8')

devicendarray_from_buffer = np.asarray(buffer_region)  # called in cudf.buffer.py->rmm.auto_device
assert devicendarray_from_buffer.dtype == np.dtype('<i8')
```

**Comment by quasiben** [comment]
@kkraus14 and I chatted for a bit and we need to resolve a few things:

1. When data is transferred with UCX how does RMM come to "own" the GPU backed ?  Should we copy the data for RMM ?
2. When UCX receives data is it responsible for allocating host/device memory ?
3. How does UCX clean up data ?

cc @Akshay-Venkatesh

**Comment by quasiben** [comment]
Getting close to a working solution but copying a DeviceNDArray built from a ucp.BufferRegion is giving me some trouble:

```python
ucp_device_array = numba.cuda.as_cuda_array(buffer)
cudautils.copy_array(ucp_device_array)
```

results in:
> *** numba.cuda.cudadrv.driver.CudaAPIError: [1] Call to cuMemcpyDtoD results in CUDA_ERROR_INVALID_VALUE

**Comment by quasiben** [comment]
I suspect the above error is because of the previous questions -- is UCX allocating resource on the GPU when it receives data

**Comment by Akshay-Venkatesh** [comment]
@quasiben Yes. If `ucp.recv_obj()` is called then a new buffer on device is allocated if the receive is specified as a cuda receive (else host buffer is allocated). This is an area that needs improvement in the sense that the receive isn't suballocating from say RMM and also because the allocation falls in the critical path.

**Comment by kkraus14** [comment]
@quasiben is this resolved or is this being worked on in #2378?

**Comment by quasiben** [comment]
Yes, it will be.  Testing now

**Comment by kkraus14** [comment]
Resolved by #2378
